### DISCUSSION TITLE
Preferred method for adding a new ecs-agent configuration parameter

### DISCUSSION THREAD

**Discussion by mello7tre** [bottlerocket-os/bottlerocket#1731] [Q&A]
Hi, looking at bottlerocket sources i have seen that a new ecs-agent configuration parameter, assuming that can be defined using json configuration, can by added changing the code in `sources/api/ecs-settings-applier/src/ecs.rs` (struct ECSConfig) or by using the environment variable file `packages/ecs-agent/ecs.config`.
(for both method other files need to be changed too, but i want to focus on the method specific one)

Which is the preferred one ? Thanks.

**Comment by zmrow** [comment]
Just to clarify, are you thinking of adding a new setting for the ECS agent that can be configured via user data?  Something like `settings.ecs.<your setting>`?

**Reply by mello7tre** [reply]
i would like to have the settings ECS_IMAGE_PULL_BEHAVIOR.
And if it is not already scheduled to be added, maybe i can do it.

**Reply by tjkirch** [reply]
If you're talking about adding a setting that's available to any Bottlerocket user, not just to your custom builds of Bottlerocket, then you'd change the ECS data model.  There's `ECSConfig`, as you saw, for settings that wind up in the ECS Agent config file.  Then, the `ECSSettings` structure in `sources/models/src/lib.rs`, which is where `ecs.rs` gets the data from, and is what allows users to specify a setting like `settings.ecs.image-pull-behavior`.  You'd also need to add a migration, which allows existing users to upgrade and downgrade Bottlerocket and have the new setting handled automatically during the transition.

If you're interested in implementing that yourself, we have a lot of examples in pull requests where we've added the settings we have today.  A recent example is for the [hostname setting](https://github.com/bottlerocket-os/bottlerocket/pull/1664/).  There were commits to handle the [modeled type](https://github.com/bottlerocket-os/bottlerocket/pull/1664/commits/5fe7fa5ec75f0279f449e4e66b55656b68767b45), which ensures the value of the setting is valid; you'd do something similar for the valid options of image pull behavior, though perhaps with an enum [like this change](https://github.com/bottlerocket-os/bottlerocket/pull/1659/commits/d0de57c8f5b44dcdba9a5f6e61190e4780deaaed) since there are only a few valid options.  The migration would look [something like this](https://github.com/bottlerocket-os/bottlerocket/pull/1659/commits/8af98dccc46df191f8dcba774d0a7e190abee8c2).  And then a [README update](https://github.com/bottlerocket-os/bottlerocket/pull/1664/commits/bda731a22c4814a3a7ebcc3442d02399c98faf81). :)

If you're just talking about adding that setting to your own custom build, then you can change the `ecs.config` file however you like, and not worry about any of the above.

**Reply by tjkirch** [reply]
(Updated the link to a migration example; you'd be adding a single setting, not a whole prefix, so #1659 is simpler than #1664.)

**Reply by mello7tre** [reply]
Thanks @tjkirch, i have already figured out all this, because i had a deep look at the sources to add settings (not ecs) required by a new program (https://github.com/bottlerocket-os/bottlerocket/pull/1728).
But your clear explanation is welcome.

Coming back to my first answer, assuming that i want to create a PR to add a settings for any Bottlerocket users, which method do your prefer:
* sources/api/ecs-settings-applier/src/ecs.rs
adding a _key_ to the `struct ECSConfig` (using proper parameter validation)
* packages/ecs-agent/ecs.config [environment]
adding something like:
ECS_IMAGE_PULL_BEHAVIOR="{{settings.ecs.image-pull-behavior}}"

**Reply by tjkirch** [reply]
Ah, I'm sorry I misunderstood the question.  The short answer is that anything that can go into `ECSConfig` in `ecs.rs` should go there.  The available attributes are defined in [this schema](https://github.com/aws/amazon-ecs-agent/blob/master/agent/config/types.go#L31) for the ecs-agent, which does include pull behavior.  Values in this file are passed directly to the agent.  If a setting isn't available in the schema, it goes in `ecs.config` and is passed to the agent using an environment variable.  (Thanks to @samuelkarp for [background](https://github.com/aws/amazon-ecs-agent/issues/409#issuecomment-219768529) here!)
